### DISCUSSION TITLE
MWL Feed from HL7 OMI Message: only the first IPC segment is considered

### DISCUSSION THREAD

**Discussion by gunterze** [dcm4che/dcm4chee-arc-light#3950] [General]
```console
$ hl7snd -c localhost:2575 /tmp/OMI^O23-with-two-IPC.hl7
16:56:58.155 INFO  - Initiate connection from 0.0.0.0/0.0.0.0:0 to localhost:2575
16:56:58.161 INFO  - Established connection Socket[addr=localhost/127.0.0.1,port=2575,localport=50787]
16:56:58.163 INFO  - Socket[addr=localhost/127.0.0.1,port=2575,localport=50787] << MSH|^~\&|MESA_OF|XYZ_RADIOLOGY|MESA_IM|XYZ_IMAGE_MANAGER|20220324193155||OMI^O23^OMI_O23|1001125|P|2.5.1||||||8859/1||
16:56:58.163 DEBUG - Socket[addr=localhost/127.0.0.1,port=2575,localport=50787] << MSH|^~\&|MESA_OF|XYZ_RADIOLOGY|MESA_IM|XYZ_IMAGE_MANAGER|20220324193155||OMI^O23^OMI_O23|1001125|P|2.5.1||||||8859/1||
PID|||PID_1^^^ADT1||Smith^Lucy^Mark||20141014|F|||PatientAddress|||||||AdmissionIDFromPatientAccountNo||||||||||||||||||||
NTE|||PatientComments||
PV1||E||||||^ReferringPhysicianFN^ReferringPhysicianGN^ReferringPhysicianMN^^DR^Md|||||||B6||||AdmissionIDFromVisitNo||||||||||||||||||||||||||||||||V|
ORC|NW|PlacerOrderNumberImagingServiceRequest|FillerOrderNumberImagingServiceRequest||SC||^^^^^RequestedProcedurePriority||||||||||HospID_1^Hospital1^CCN|||||InstitutionAddress|||||
TQ1|||||||200008161510||
OBR||PlacerOrderNumberImagingServiceRequest|FillerOrderNumberImagingServiceRequest|10637-7^Microscopic Observation^LN||||||||PatientState|MedicalAlerts|||^RequestingPhysicianFN^RequestingPhysicianGN^RequestingPhysicianMN^^DR^Md||||||||||||||PatientTransportArrangements|D1304^Rheumatoid, other arthritis with motor score from 12 –35.,comorbidity in tier 3^HIPPS|||ScheduledPerformingPhysicianNameFN^ScheduledPerformingPhysicianNameGN^ScheduledPerformingPhysicianNameMN^^DR^Md||||||||||10637-7^Microscopic Observation^LN|
OBX|1|ST|^Body Weight||65|kg|||||F||||||
OBX|2|ST|^Body Height||1.65|m|||||F||||||
NTE|||RequestedProcedureComments||||
IPC|AccessionNumber|RequestedProcID|1.2.392.200036.9125.0.198811291108.7|ScheduledProcStepID|CT|10637-7^Microscopic Observation^LN|ScheduledStationName|ScheduledProcStepLocation|ScheduledStationAET1~ScheduledStationAET2
IPC|AccessionNumber|RequestedProcID|1.2.392.200036.9125.0.198811291108.7|ScheduledProcStepID2|CT|10637-7^Microscopic Observation^LN|ScheduledStationName|ScheduledProcStepLocation|ScheduledStationAET1~ScheduledStationAET2

16:56:58.200 INFO  - Socket[addr=localhost/127.0.0.1,port=2575,localport=50787] >> MSH|^~\&|MESA_IM|XYZ_IMAGE_MANAGER|MESA_OF|XYZ_RADIOLOGY|20230203165658.164||ACK^O23^ACK|2052178904|P|2.5.1||||||8859/1||
16:56:58.200 DEBUG - Socket[addr=localhost/127.0.0.1,port=2575,localport=50787] >> MSH|^~\&|MESA_IM|XYZ_IMAGE_MANAGER|MESA_OF|XYZ_RADIOLOGY|20230203165658.164||ACK^O23^ACK|2052178904|P|2.5.1||||||8859/1||
MSA|AA|1001125|

16:56:58.200 INFO  - Close connection Socket[addr=localhost/127.0.0.1,port=2575,localport=50787]
```
results in only **one** created MWL item with the Scheduled Procedure ID `ScheduledProcStepID` from the first `IPC` segment.
```console
$ curl http://localhost:8080/dcm4chee-arc/aets/DCM4CHEE/rs/mwlitems?StudyInstanceUID=1.2.392.200036.9125.0.198811291108.7 | jq | grep -A4 00400009
          "00400009": {
            "vr": "SH",
            "Value": [
              "ScheduledProcStepID"
            ]
```
Seems XSLT stylesheet [hl7-order2dcm.xsl](https://github.com/dcm4che/dcm4chee-arc-light/blob/master/dcm4chee-arc-conf-data/src/main/resources/hl7-order2dcm.xsl) does not handle multiple IPCs correctly - as it does for multiple ORC segments in HL7 ORM messages.

[OMI^O23-with-two-IPC.hl7.zip](https://github.com/dcm4che/dcm4chee-arc-light/files/10580628/OMI.O23-with-two-IPC.hl7.zip)

According [IHE Radiology Technical Framework, Volume 2 (TF-2), p. 118](https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=118):

> One ORC-TQ1-OBR-IPC segment group shall correspond to each Requested Procedure. If a
Requested Procedure is comprised of multiple Scheduled Procedure Steps and/or if a Scheduled3085
Procedure Step is comprised of multiple Protocol Codes, each applicable Scheduled Procedure
Step / Protocol Code combination shall be included in a separate IPC segment following the
ORC-TQ1-OBR segment group that contains the Requested Procedure.

**Comment by vrindanayak** [comment]
Archive currently supports multiple ORC-TQ1-OBR-IPC group => to be translated to multiple MWLs / Scheduled Procedure Steps

````
MSH|^~\&|MESA_OF|XYZ_RADIOLOGY|MESA_IM|XYZ_IMAGE_MANAGER|20221212201111||ORM^O01|104106|P|2.3.1||||||||
PID|||5830401^^^ADT11||BLUE1^ERICA1||18980502|F||BL|15 PRINCETON^^ST. LOUIS^MO^63130|||||||20983040|||||||||||||||||||||
PV1||E|ED||||3456^ROSZEK^JEANETTE^G^^DR|5202^SCHULTZ^KEVIN^G^^DR|||||||||||V3041^^^ADT1|||||||||||||||||||||||||200008101130||||||||
ORC|NW|A104Z^MESA_ORDPLC|B104Z^MESA_ORDFIL||SC||||200008101330|^ROSEWOOD^RANDOLPH||7202^ESSELMAN^HENRY^G^^DR||(314)555-1212|200008101330||922229-10^IHE-RAD^IHE-CODE-231||
TQ1|1|1|once||||20000816151515|
OBR|1|A104Z^MESA_ORDPLC|B104Z^MESA_ORDFIL|P4^Procedure 4^ERL_MESA^X4A_A1^SP Action Item X4A_A1^DSS_MESA|||||||||xxx||Radiology^^^^R|7202^ESSELMAN^HENRY^G^^DR||ACCESSION_NUMBER|REQUESTED_PROCEDURE_ID|SCHEDULED_PROCEDURE_STEP_ID_1||||MR||||||WALK|||||||||||A|||
IPC|AI|REQUESTED_PROCEDURE_ID|1.113654.3.104.1|SCHEDULED_PROCEDURE_STEP_ID_1|MR|PROTOCOL_CODE1|||AET1~AET2|
ORC|NW|A104Z^MESA_ORDPLC|B104Z^MESA_ORDFIL||SC||||200008101330|^ROSEWOOD^RANDOLPH||7202^ESSELMAN^HENRY^G^^DR||(314)555-1212|200008101330||922229-10^IHE-RAD^IHE-CODE-231||
TQ1|2|1|once||||20000817151515|
OBR|2|A104Z^MESA_ORDPLC|B104Z^MESA_ORDFIL|P4^Procedure 4^ERL_MESA^X4B_A1^SP Action Item X4B_A1^DSS_MESA|||||||||xxx||Radiology^^^^R|7202^ESSELMAN^HENRY^G^^DR||ACCESSION_NUMBER|REQUESTED_PROCEDURE_ID|SCHEDULED_PROCEDURE_STEP_ID_2||||CT||||||WALK|||||||||||A|||
IPC|AI|REQUESTED_PROCEDURE_ID|1.113654.3.104.1|SCHEDULED_PROCEDURE_STEP_ID_2|US|PROTOCOL_CODE2|||AET3~AET4|
ORC|NW|A104Z^MESA_ORDPLC|B104Z^MESA_ORDFIL||SC||||200008101330|^ROSEWOOD^RANDOLPH||7202^ESSELMAN^HENRY^G^^DR||(314)555-1212|200008101330||922229-10^IHE-RAD^IHE-CODE-231||
TQ1|3|1|once||||20000818151515|
OBR|3|A104Z^MESA_ORDPLC|B104Z^MESA_ORDFIL|P4^Procedure 4^ERL_MESA^X4B_A2^SP Action Item X4B_A2^DSS_MESA|||||||||xxx||Radiology^^^^R|7202^ESSELMAN^HENRY^G^^DR||ACCESSION_NUMBER|REQUESTED_PROCEDURE_ID|SCHEDULED_PROCEDURE_STEP_ID_3||||CT||||||WALK|||||||||||A|||
IPC|AI|REQUESTED_PROCEDURE_ID|1.113654.3.104.1|SCHEDULED_PROCEDURE_STEP_ID_3|CT|PROTOCOL_CODE3|||AET5~AET6|
````

![image](https://user-images.githubusercontent.com/15785031/216692288-aaca3d75-be4a-41c2-801d-6726d4f8e621.png)

````
curl http://localhost:8880/dcm4chee-arc/aets/WORKLIST/rs/mwlitems?StudyInstanceUID=1.113654.3.104.1 | jq | grep -A4 00400009
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4597    0  4597    0     0   122k      0 --:--:-- --:--:-- --:--:--  124k
          "00400009": {
            "vr": "SH",
            "Value": [
              "SCHEDULED_PROCEDURE_STEP_ID_1"
            ]
--
          "00400009": {
            "vr": "SH",
            "Value": [
              "SCHEDULED_PROCEDURE_STEP_ID_2"
            ]
--
          "00400009": {
            "vr": "SH",
            "Value": [
              "SCHEDULED_PROCEDURE_STEP_ID_3"
            ]
````

See https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=118
> One ORC-TQ1-OBR-IPC segment group shall correspond to each Requested Procedure. If a Requested Procedure is comprised of multiple Scheduled Procedure Steps and/or if a Scheduled Procedure Step is comprised of multiple Protocol Codes, each applicable Scheduled Procedure Step / Protocol Code combination shall be included in a separate IPC segment following the ORC-TQ1-OBR segment group that contains the Requested Procedure.

**Comment by gunterze** [comment]
I read
> One ORC-TQ1-OBR-IPC segment group shall correspond to each Requested Procedure. If a Requested Procedure is comprised of multiple Scheduled Procedure Steps and/or if a Scheduled Procedure Step is comprised of multiple Protocol Codes, each applicable Scheduled Procedure Step / Protocol Code combination shall be included in a separate IPC segment following the ORC-TQ1-OBR segment group that contains the Requested Procedure.

that there should be only **one** ORC-TQ1-OBR segment group in one OMI - which itself correspond to **one** Requested Procedure - followed by `[1..*]` IPC segment corresponding to each Procedure Step of the requested Procedure.

In opposite to the segment list for the ORM message at https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=109, **no** repeating group of segments is specified in the segment list for the OMI message at https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=115 and the cardinality of the segments TQ1 and OBR are restricted to `[1..1]`. Use of multiple ORC segments is not mentioned in _4.4.4.1.2.2.5 ORC Segment (HL7 v2.5.1)_, so I rather guess the cardinality `[1..*]` for ORC in the segment list is a typo.

And

> The Department System Scheduler/Order Filler shall send as many OMI messages as there are Requested Procedures identified to fill a single order.

together with

> One ORC-TQ1-OBR-IPC segment group shall correspond to each Requested Procedure. ...

concludes, that there is only **one** ORC-TQ1-OBR in **one** OMI message.

In any case, there is room for improving the readability of IHE RAD TF-2, 4.4.4.1.2.2 Message Semantics (HL7 v2.5.1).

**Reply by vrindanayak** [reply]
Ok, understood and agreed. However, as you've mentioned earlier in an internal chat, we need to get it clarified from IHE Technical Committee on how to get the information about Scheduled Procedure Step Start Date Time for multiple Scheduled Procedure Steps, as this is retrieved from [TQ1.7](https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/TQ1.7), according to and as required for [RAD-4 Procedure Scheduled HL7 v2.5.1  IHE Profile – TQ1 Segment](https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=119)

**Reply by gunterze** [reply]
The intention of IHE Transaction [Procedure Scheduled \[RAD-4\]](https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=105) is defined as

> This message serves as a trigger event for the Image Manager and the Report Manager,
informing it to obtain necessary information and apply rules to ensure the availability of relevant
information to the end user. The Image Manager and the Report Manager may need the
information to create the Requested Procedure context for its purposes.

Particularly, none of the discussed use cases under [4.4.4.2 Expected Actions](https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_TF_Vol2.pdf#page=123) mention the feeding of a DICOM MWL SCP by such messages. So IHE may just declare that the single Start Date/Time passed in TQ1.7 shall be equal or before the  Start Date Time of the first Scheduled Procedure Step of the Requested Procedure. So we would need to define a proprietary solution, if just using a common Start Date Time for all  Scheduled Procedure Step belonging to one Requested Procedure is not appropriate for dcm4chee-arc users. But I prefer to wait for such user request, before already discussing possible technical solutions.

**Comment by vrindanayak** [comment]
Solved by https://github.com/dcm4che/dcm4chee-arc-light/issues/4526
